aptos-move/framework/supra-framework/sources/multisig_voting.spec.move

```diff
@@ -0,0 +1,427 @@
+    }
+
+    spec create_proposal_v2 {
+        use supra_framework::chain_status;
+        //TODO: Remove pragma aborts_if_is_partial;
+        pragma aborts_if_is_partial = true;
+        requires chain_status::is_operating();
+        include CreateProposalAbortsIfAndEnsures<ProposalType>;
+        // property 1: Verify the proposal_id of the newly created proposal.
+        ensures result == old(global<VotingForum<ProposalType>>(voting_forum_address)).next_proposal_id;
+    }
+
+    spec vote {
+        use supra_framework::chain_status;
+        // Ensures existence of Timestamp
+        requires chain_status::is_operating();
+        //TODO: Remove pragma aborts_if_is_partial;
+        pragma aborts_if_is_partial = true;
+        // property 2: While voting, it ensures that only the governance module that defines ProposalType may initiate voting
+        // and that the proposal under vote exists in the specified voting forum.
+        /// [high-level-req-2]
+        aborts_if !exists<VotingForum<ProposalType>>(voting_forum_address);
+        let voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let proposal = table::spec_get(voting_forum.proposals, proposal_id);
+        // Getting proposal from voting forum might fail because of non-exist id
+        aborts_if !table::spec_contains(voting_forum.proposals, proposal_id);
+        aborts_if is_voting_period_over(proposal);
+        aborts_if proposal.is_resolved;
+        aborts_if !exists<timestamp::CurrentTimeMicroseconds>(@supra_framework);
+        // Assert this proposal is single-step, or if the proposal is multi-step, it is not in execution yet.
+        aborts_if !std::string::spec_internal_check_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        let execution_key = std::string::spec_utf8(IS_MULTI_STEP_PROPOSAL_IN_EXECUTION_KEY);
+        aborts_if simple_map::spec_contains_key(proposal.metadata, execution_key) &&
+            simple_map::spec_get(proposal.metadata, execution_key) != std::bcs::serialize(false);
+        // aborts_if if (should_pass) { proposal.yes_votes + num_votes > MAX_U128 } else { proposal.no_votes + num_votes > MAX_U128 };
+
+        aborts_if !std::string::spec_internal_check_utf8(RESOLVABLE_TIME_METADATA_KEY);
+
+        let post post_voting_forum = global<VotingForum<ProposalType>>(voting_forum_address);
+        let post post_proposal = table::spec_get(post_voting_forum.proposals, proposal_id);
+        // ensures if (should_pass) {
+        //     post_proposal.yes_votes == proposal.yes_votes + num_votes
+        // } else {
+        //     post_proposal.no_votes == proposal.no_votes + num_votes
+        // };
+        let timestamp_secs_bytes = std::bcs::serialize(timestamp::spec_now_seconds());
+        let key = std::string::spec_utf8(RESOLVABLE_TIME_METADATA_KEY);
+        ensures simple_map::spec_get(post_proposal.metadata, key) == timestamp_secs_bytes;
```

Comment:
This property only ensures that time is recorded. Other properties we should aim at adding are

- vote by the voter is recorded (if not already voted)
- vote by the voter is recorded (if vote is flipped)
- total yes/no votes should tally (increase/decrease appropriately)

Reply:
Added in [e65045b](https://github.com/Entropy-Foundation/aptos-core/pull/63/commits/e65045bb766d6d3a5649667fae8a40b5b8dd79ef)